runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/net/newnet/Url.kt

```diff
@@ -0,0 +1,264 @@
+                }
+            }
+
+            scanner.ifStartsWith("?") {
+                scanner.upToOrEnd("#") {
+                    queryParameters { parseEncoded(it) }
+                }
+            }
+
+            scanner.ifStartsWith("#") {
+                scanner.upToOrEnd { fragmentEncoded = it }
+            }
+        }
+    }
+
+    init {
+        require(port in 1..65535) { "Given port $port is not in required range [1, 65535]" }
+    }
+
+    /**
+     * Copy the properties of this [Url] instance into a new [Builder] object. Any changes to the builder *will not*
+     * affect this instance.
+     */
+    public fun toBuilder(): Builder = Builder(this)
+
+    /**
+     * Returns the encoded string representation of this URL
+     */
+    override fun toString(): String = buildString {
+        append(scheme.protocolName)
+        append("://")
+        append(host.toUrlString())
+        if (port != scheme.defaultPort) {
+            append(":")
+            append(port)
+        }
+        append(path)
+        queryParameters?.let {
+            append('?')
+            append(it)
+        }
+        fragment?.let {
+            append('#')
+            append(it.encoded)
+        }
+    }
+
+    /**
+     * A mutable builder used to construct [Url] instances
+     */
+    public class Builder internal constructor(url: Url?) {
+        /**
+         * Initialize an empty [Url] builder
+         */
+        public constructor() : this(null)
+
+        // Simple fields
+
+        /**
+         * The wire protocol (e.g., http, https, etc.)
+         */
+        public var scheme: Scheme = url?.scheme ?: Scheme.HTTPS
+
+        /**
+         * The [Host] for the URL
+         */
+        public var host: Host = url?.host ?: Host.Domain("")
+
+        /**
+         * The remote port number for the URL (e.g., TCP port)
+         */
+        public var port: Int? = url?.port
+
+        // Path
+
+        private val path: UrlPath.Builder = url?.path?.toBuilder() ?: UrlPath.Builder()
+
+        /**
+         * Update the [UrlPath] of this URL via a DSL builder block
+         * @param block The code to apply to the [UrlPath] builder
+         */
+        public fun path(block: UrlPath.Builder.() -> Unit) {
+            path.apply(block)
+        }
+
+        /**
+         * Get or set the URL path as a **decoded** string
+         */
+        public var pathDecoded: String
+            get() = path.asDecoded()
+            set(value) { path.parseDecoded(value) }
+
+        /**
+         * Get or set the URL path as an **encoded** string
+         */
+        public var pathEncoded: String
+            get() = path.asEncoded()
+            set(value) { path.parseEncoded(value) }
+
+        // Query parameters
+
+        private var queryParameters = url?.queryParameters?.toBuilder()
+
+        /**
+         * Remove all query parameters from this URL
+         */
+        public fun clearQueryParameters() {
```

Comment:
Previously we handled this with the [forceQuery](https://github.com/awslabs/smithy-kotlin/blob/c15956eeddaed8dedd6640ac60e7d291416a92be/runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/net/Url.kt#L87) parameter on the builder. I'm not suggesting to ignore this but it's not something we've really needed or used much so can probably be moved to a boolean flag like that (either under `Url.Builder` directly or as part of `QueryParameters[.Builder])`). 

Mostly I think it's an ergonomics thing and IMHO the non-nullably empty version of this makes it easier to consume and composes better (whereas making it nullable hoists a bunch of concerns of it's parts into the `Url.Builder` directly rather than letting the individual part deal with it).